pwndbg/color/disasm.py

```diff
@@ -22,67 +23,123 @@
-        if is_branch and not ins.target:
-            asm = f"{asm} <{ins.symbol}>"
-
-            # XXX: not sure when this ever happens
-            asm += "<-- file a pwndbg bug for this"
-        else:
-            inlined_sym = asm.replace(hex(ins.symbol_addr), ins.symbol)
+    is_call_or_jump = ins.groups_set & capstone_branch_groups
 
-            # display symbol as mem text if no inline replacement was made
-            mem_text = ins.symbol if inlined_sym == asm else None
-            asm = f"{ljust_colored(inlined_sym, 36)} <{M.get(ins.symbol_addr, mem_text)}>"
-
-    # Style the instruction mnemonic if it's a branch instruction.
-    if is_branch:
+    # Style the instruction mnemonic if it's a call/jump instruction.
+    if is_call_or_jump:
         asm = asm.replace(ins.mnemonic, c.branch(ins.mnemonic), 1)
 
     # If we know the conditional is taken, mark it as taken.
-    if ins.condition is None:
-        asm = "  " + asm
-    elif ins.condition:
+    if ins.condition == InstructionCondition.TRUE or ins.is_conditional_jump_taken:
         asm = on("✔ ") + asm
     else:
         asm = "  " + asm
 
     return asm
+
+
+# To making the padding visually nicer, the following padding scheme is used for annotations:
+# All instructions in a group will have the same amount of left-adjusting spaces, so they are aligned.
+# A group is defined as a sequence of instructions surrounded by instructions that can change the instruction pointer.
+def instructions_and_padding(instructions: list[PwndbgInstruction]) -> list[str]:
+    assembly = [one_instruction(i) for i in instructions]
+
+    result: list[str] = []
+
+    MIN_SPACING = 5
+
+    WHITESPACE_LIMIT = 20
+
+    cur_padding_len = None
+
+    # Stores intermediate padding results so we can do a final pass to clean up edges and jagged parts
+    # None if padding doesn't apply to the instruction
+    paddings: list[int | None] = []
+
+    # Used for padding. List of groups.
+    # Each group is a list of index into paddings list
+    groups: list[list[int]] = []
+
+    current_group: list[int] = []
+
+    for i, (ins, asm) in enumerate(zip(instructions, assembly)):
+        if ins.can_change_instruction_pointer:
+            sym = ins.target_string
+
+            asm = f"{ljust_colored(asm, 36)} <{sym}>"
+
+            paddings.append(None)
+            if current_group:
+                groups.append(current_group)
+                current_group = []
+        else:
+            # Padding the string for a nicer output
+            # This path calculates the padding for each instruction - even if there we don't have annotations for it.
+            # This allows groups to have uniform padding, even if some of the instructions don't have annotations
+            current_group.append(i)
+
+            raw_len = len(strip(asm))
+
+            if cur_padding_len is None:
+                cur_padding_len = raw_len + MIN_SPACING
+
+            if cur_padding_len - raw_len < MIN_SPACING:
```

Comment:
Should this be 'elif' or not really?
```suggestion
            elif cur_padding_len - raw_len < MIN_SPACING:
```

Reply:
Elif works here - changed.